Runtime/Services/BacktraceSession.cs

```diff
@@ -0,0 +1,317 @@
+        public BacktraceSession(
+            AttributeProvider attributeProvider,
+            string uploadUrl,
+            long timeIntervalInMs,
+            uint maximumNumberOfEventsInStore)
+        {
+            SubmissionUrl = uploadUrl;
+            _attributeProvider = attributeProvider;
+            _maximumNumberOfEventsInStore = maximumNumberOfEventsInStore;
+            _timeIntervalInMs = timeIntervalInMs;
+            RequestHandler = new BacktraceHttpClient();
+        }
+
+        /// <summary>
+        /// Backtrace session update interval method used by Backtrace client to update session events time
+        /// </summary>
+        /// <param name="time">Current game time</param>
+        public void Tick(long time)
+        {
+            if (_timeIntervalInMs == 0)
+            {
+                return;
+            }
+            lock (_object)
+            {
+                var intervalUpdate = (time - _lastUpdateTime) >= _timeIntervalInMs;
+                var reachedEventLimit = _maximumNumberOfEventsInStore == Count() && _maximumNumberOfEventsInStore != 0;
+                if (intervalUpdate == false && reachedEventLimit == false)
+                {
+                    // nothing more to update
+                    return;
+                }
+                _lastUpdateTime = time;
+            }
+            Send();
+        }
+
+        /// <summary>
+        /// Force BacktraceSession to Send data to Backtrace
+        /// </summary>
+        public void Send()
+        {
+            Send(numberOfRetries: 3);
+        }
+
+        /// <summary>
+        /// Force BacktraceSession to Send data to Backtrace with retry settings
+        /// </summary>
+        public void Send(int numberOfRetries = DefaultNumberOfRetries)
+        {
+            if (numberOfRetries == 0)
+            {
+                Debug.LogWarning("Backtrace Session: Cannot send session data to Backtrace due to submission issue.");
+                CleanEvents();
+                return;
+            }
+            if (Count() == 0)
+            {
+                return;
+            }
+            // create a copy of events 
+            var uniqueEvents = UniqueEvents.ToArray();
+            var sessionEvents = SessionEvents.ToArray();
+            var payload = CreateJsonPayload(uniqueEvents, sessionEvents);
+
+            // cleanup existing copy of events
+            CleanEvents();
+
+            // submit data to Backtrace
+            RequestHandler.Post(SubmissionUrl, payload, (long statusCode, bool httpError, string response) =>
+            {
+                if (statusCode == 200)
+                {
+                    OnRequestCompleted();
+                }
+                else
+                {
+                    OnRequestFailure(uniqueEvents, sessionEvents);
+                    // assume that we should try to retry request on 503

```

Comment:
Is this comment referring to logic that's going to be in the other prs?  it looks like the current logic retries for any status code that's not exactly 200 (e.g. other 2xx responses also retry, etc).

Reply:
Based on our Slack thread available here: https://backtrace.slack.com/archives/C01NYBYN7MY/p1618248208041800 we wanted to:
1. avoid retries (Samy)
2. retry on 503

I thought retry on at least 503 give us something. 

I think we will review retry logic later so this part of the pull request might change after first round of feedback from our customers/tech services/ other